### ISSUE TITLE
Smart copy content of latexfiles and sources

### ISSUE THREAD

**Issue by nowox** [sphinx-doc/sphinx#2510]
Currently, sphinx recopies all the LaTeX files (figures/texfiles/extras) as showed below: 

```
    def finish(self):
        # copy image files
        if self.images:
            self.info(bold('copying images...'), nonl=1)
            for src, dest in iteritems(self.images):
                self.info(' '+src, nonl=1)
                copyfile(path.join(self.srcdir, src),
                         path.join(self.outdir, dest))
            self.info()

        # copy TeX support files from texinputs
        self.info(bold('copying TeX support files...'))
        staticdirname = path.join(package_dir, 'texinputs')
        for filename in os.listdir(staticdirname):
            if not filename.startswith('.'):
                copyfile(path.join(staticdirname, filename),
                         path.join(self.outdir, filename))

        # copy additional files
        if self.config.latex_additional_files:
            self.info(bold('copying additional files...'), nonl=1)
            for filename in self.config.latex_additional_files:
                self.info(' '+filename, nonl=1)
                copyfile(path.join(self.confdir, filename),
                         path.join(self.outdir, path.basename(filename)))
            self.info()
```

I imagine a more sophisticated approach with a `copy` function that only copy a file when the content differs.

```
    def copy_smart(self, src, dst, ignore = lambda: false):
        """Copy file from source to destination if content differs.

        This method copies the `src` file into `dst` only if:

        - Destination file does not exists
        - Destination file exists, but differ from the source file

        If source path `src` is a list, `dst` must be a directory.

        Args:
            src (str/list): Source path (absolute or relative path)
            dst (str): Destination path ( absolute or relative path)
            ignore (lambda): Ignore file if lambda returns true
        Note:
            `rsync --checksum` could be used to increase performances.

        """
        if type(src) is not list:
            src = [src]

        if os.path.isdir(dst): 
            dst = map(lambda x: path.join(dst, path.basename(x)), src)

        for s, d in zip(src, dst):
            if ignore(s):
                continue
            if not os.path.exists(d) or not filecmp.cmp(s, d):
                self.info(' %s -> %s' % (os.path.relpath(s), os.path.relpath(d)))
                copyfile(s, d)

    def finish(self):
        # copy image files
        if self.images:
            self.info(bold('copying images...'))
            self.copy_smart(path.join(self.srcdir, self.images.keys()),
                            path.join(self.outdir, self.images.values()))
            self.info()

        # copy TeX support files from texinputs
        self.info(bold('copying TeX support files...'))
        staticdirname = path.join(package_dir, 'texinputs')
        src = [path.join(staticdirname, x) for x in os.listdir(staticdirname)]
        self.copy_smart(src, self.outdir, ignore = lambda x: x.startswith('.'))

        # copy additional files
        if self.config.latex_additional_files:
            self.info(bold('copying additional files...'))
            self.copy_smart(self.config.latex_additional_files, self.outdir)
            self.info()

        # the logo is handled differently
        if self.config.latex_logo:
            logobase = path.basename(self.config.latex_logo)
            logotarget = path.join(self.outdir, logobase)
            if not path.isfile(path.join(self.confdir, self.config.latex_logo)):
                raise SphinxError('logo file %r does not exist' % self.config.latex_logo)
            elif not path.isfile(logotarget):
                self.copy_smart(path.join(self.confdir, self.config.latex_logo), logotarget)
        self.info('done')    
```

I am not sure about the name of this method and whether to put it on the LaTeX builder class.

**Comment by tk0miya** [comment]
MEMO: #1916 is similar to this.

**Comment by tk0miya** [comment]
Fixed in #2754
